Agent can't access control plane

**Describe the bug**
I have installed the control plane (Version 1.3.0) in k8s via the helm chart, and publish it via project contour Ingress controller, witch works only with namebased virtual hosting.
I can't make the agent on a separate vm to work with the control plane, cause it seems that the agent try to make the API requests using the IP address instead of the hostname. Because of this the namebased virtualhosting in project contour doesnt work and the agent cant reach the control plane.

**To Reproduce**
- Install control plane from helm chart
- Publish it with contour httpproxy:
```yaml
apiVersion: projectcontour.io/v1
kind: HTTPProxy
metadata:
  name: treatmapper-tls
spec:
  virtualhost:
    fqdn: treatmapper.demo
    tls:
      passthrough: true
  tcpproxy:
    services:
      - name: deepfence-router
        port: 443
  routes:
    - services:
      - name: deepfence-router
        port: 80
```
- Start a agent on a separete VM with:
```bash
docker run -dit --cpus=".2" --name=deepfence-agent --restart on-failure --pid=host --net=host \
  --privileged=true -v /sys/kernel/debug:/sys/kernel/debug:rw -v /var/log/fenced \
  -v /var/run/docker.sock:/var/run/docker.sock -v /:/fenced/mnt/host/:ro \
  -e USER_DEFINED_TAGS="" -e MGMT_CONSOLE_URL="treatmapper.demo" -e MGMT_CONSOLE_PORT="443" \
  -e DEEPFENCE_KEY="KEY" -e LOG_LEVEL="debug"\
  deepfenceio/deepfence_agent_ce:1.3.0
```

**Expected behavior**
The agent can communicate with control plane, using the hostname in requests

**Screenshots**
I can see this in agent logs (inside the container in /var/log/fenced/discovery.logfile):
```log
<probe> DEBU: 2022/07/08 06:36:53.613158 background /proc reader: full pass took 264.715148ms
<probe> DEBU: 2022/07/08 06:36:53.613208 background /proc reader: new rate limit period 500ms
<probe> ERRO: 2022/07/08 06:37:00.107353 Error fetching app details: Get "https://10.121.0.30:443/topology-api": read tcp 10.121.0.127:2428->10.121.0.30:443: read: connection reset by peer
<probe> DEBU: 2022/07/08 06:37:03.247119 docker container: collecting stats for 8795da4c7741d5b2b0fa738ef235463eaa5f0e7a6afcc2d1277062b0787de182
<probe> DEBU: 2022/07/08 06:37:03.364350 docker container: stopped collecting stats for 8795da4c7741d5b2b0fa738ef235463eaa5f0e7a6afcc2d1277062b0787de182
```
Where 10.121.0.30 is the IP of DNS treatmapper.demo.




**Components/Services affected**

- [ ] UI/Frontend
- [ ] API/Backend
- [x] Agent
- [ ] Deployment/YAMLs
- [ ] CI/CD Integration
- [ ] Other (specify)

**Additional context**

@ramanan-ravi  thanks for your reply.  
I came to install 1.3.0 cause in 1.3.1 i have another problem in the same configuration: After i start the sensors pods in the same cluster, as the control plane, the deepfence-topology pod crash with:
```log
panic: UnmarshalJSON shouldn't be used, use CodecDecodeSelf instead

goroutine 634 [running]:
github.com/weaveworks/scope/report.(*MetricRow).UnmarshalJSON(0xc000c80380, 0xc000b9e70e, 0x89, 0x16e73, 0x7f17b7cebd28, 0xc000c80380)
        /go/src/github.com/weaveworks/scope/report/metric_row.go:48 +0x39
encoding/json.(*decodeState).object(0xc003962dc0, 0x2110c20, 0xc000c80380, 0x199, 0x22058b4, 0xb)
        /usr/local/go/src/encoding/json/decode.go:609 +0x20a3
encoding/json.(*decodeState).value(0xc003962dc0, 0x2110c20, 0xc000c80380, 0x199, 0x2110c20, 0xc000c80380)
        /usr/local/go/src/encoding/json/decode.go:370 +0x6d
encoding/json.(*decodeState).array(0xc003962dc0, 0x1dc67e0, 0xc0041b9ba8, 0x197, 0xc003962de8, 0x5b)
        /usr/local/go/src/encoding/json/decode.go:558 +0x1aa
encoding/json.(*decodeState).value(0xc003962dc0, 0x1dc67e0, 0xc0041b9ba8, 0x197, 0x1dc67e0, 0xc0041b9ba8)
        /usr/local/go/src/encoding/json/decode.go:360 +0x105
encoding/json.(*decodeState).object(0xc003962dc0, 0x20c1640, 0xc0041b9b00, 0x199, 0xc003962de8, 0xc00462db7b)
        /usr/local/go/src/encoding/json/decode.go:765 +0x1325
encoding/json.(*decodeState).value(0xc003962dc0, 0x20c1640, 0xc0041b9b00, 0x199, 0xc0041b9b00, 0x199)
        /usr/local/go/src/encoding/json/decode.go:370 +0x6d
encoding/json.(*decodeState).object(0xc003962dc0, 0x1eea220, 0xc002ffa010, 0x16, 0xc003962de8, 0x7b)
        /usr/local/go/src/encoding/json/decode.go:765 +0x1325
encoding/json.(*decodeState).value(0xc003962dc0, 0x1eea220, 0xc002ffa010, 0x16, 0xc00462dea0, 0x7ed6cb)
        /usr/local/go/src/encoding/json/decode.go:370 +0x6d
encoding/json.(*decodeState).unmarshal(0xc003962dc0, 0x1eea220, 0xc002ffa010, 0xc003962de8, 0x0)
        /usr/local/go/src/encoding/json/decode.go:180 +0x1f4
encoding/json.Unmarshal(0xc000b9e000, 0x17581, 0x17581, 0x1eea220, 0xc002ffa010, 0x17581, 0x17581)
        /usr/local/go/src/encoding/json/decode.go:107 +0x11a
github.com/weaveworks/scope/cache.(*RedisCache).scopeTopologyFixes(0xc000c6adc0, 0xc00462e840, 0x0, 0x0)
        /go/src/github.com/weaveworks/scope/cache/cache.go:148 +0x6b8
github.com/weaveworks/scope/cache.(*RedisCache).Update(0xc000c6adc0, 0xc003f7d5f0)
        /go/src/github.com/weaveworks/scope/cache/cache.go:60 +0x65
github.com/weaveworks/scope/app.CacheTopology.func1(0xc000ce4d50, 0xc000759c00, 0x7f17d4896a80, 0xc000280380, 0x2840180, 0xc0001b4010, 0x2200e23, 0xa, 0xc000c6adc0)
        /go/src/github.com/weaveworks/scope/app/cache_topology.go:48 +0x2a8
created by github.com/weaveworks/scope/app.CacheTopology
        /go/src/github.com/weaveworks/scope/app/cache_topology.go:42 +0x732
```
After that the pods restarts and crash again:

```bash
kubectl get -A po | grep treat 
treatmapper-sensor   deepfence-agent-257mj                             1/1     Running   0          6m25s
treatmapper-sensor   deepfence-agent-9tw78                             1/1     Running   0          6m25s
treatmapper-sensor   deepfence-agent-mh5pp                             1/1     Running   0          6m25s
treatmapper-sensor   deepfence-cluster-agent-864f976dd7-2cc8f          1/1     Running   0          6m25s
treatmapper          deepfence-api-9cf647dd9-ghmvd                     1/1     Running   0          8m1s
treatmapper          deepfence-backend-58679df4bd-bq6f4                1/1     Running   0          8m1s
treatmapper          deepfence-celery-78d45f8f9b-2hrc7                 1/1     Running   0          8m1s
treatmapper          deepfence-diagnosis-986687fcd-xwmz7               1/1     Running   0          8m1s
treatmapper          deepfence-discovery-mgcvm                         1/1     Running   0          8m1s
treatmapper          deepfence-discovery-p7vb4                         1/1     Running   0          8m1s
treatmapper          deepfence-discovery-ql5v8                         1/1     Running   0          8m1s
treatmapper          deepfence-es-0                                    1/1     Running   0          8m1s
treatmapper          deepfence-fetcher-6d45f9655c-4tcjb                1/1     Running   0          8m1s
treatmapper          deepfence-internal-router-567566669f-k4fs9        1/1     Running   0          8m1s
treatmapper          deepfence-package-scanner-6c464f78cb-f7tfk        1/1     Running   0          8m1s
treatmapper          deepfence-package-scanner-6c464f78cb-l2cf9        1/1     Running   0          8m1s
treatmapper          deepfence-package-scanner-6c464f78cb-nmkh8        1/1     Running   0          8m1s
treatmapper          deepfence-redis-776fc6c767-8zvfx                  1/1     Running   0          8m1s
treatmapper          deepfence-router-67cf6c458b-8z2n4                 1/1     Running   0          8m1s
treatmapper          deepfence-router-67cf6c458b-gqx4z                 1/1     Running   0          8m1s
treatmapper          deepfence-router-67cf6c458b-j8rh2                 1/1     Running   0          8m1s
treatmapper          deepfence-secret-scanner-7cf7db96d-rmz65          1/1     Running   0          8m1s
treatmapper          deepfence-secret-scanner-7cf7db96d-vk8jm          1/1     Running   0          8m1s
treatmapper          deepfence-topology-564ff9447f-nrjdj               1/1     Running   5          8m1s
treatmapper          deepfence-ui-867dd5967b-hjqb5                     1/1     Running   0          8m1s
treatmapper          deepfence-vulnerability-mapper-7cbd7dd8cb-2nrdn   1/1     Running   0          8m1s
```
What can be the reason?

--- a/deepfence_agent/tools/apache/scope/cache/cache.go
+++ b/deepfence_agent/tools/apache/scope/cache/cache.go
@@ -1,8 +1,8 @@
 package cache
 
 import (
+	"bytes"
 	"context"
-	"encoding/json"
 	"fmt"
 	"github.com/gomodule/redigo/redis"
 	"github.com/olivere/elastic/v7"
@@ -145,7 +145,7 @@ func (r *RedisCache) scopeTopologyFixes(nodeSummaries *detailed.NodeSummaries) e
 			return err
 		}
 		var topologyPods detailed.NodeSummaries
-		err = json.Unmarshal(topologyPodsJson, &topologyPods)
+		err = CodecDecode(bytes.NewReader(topologyPodsJson), &topologyPods)
 		if err != nil {
 			return err
 		}

--- a/deepfence_agent/tools/apache/scope/cache/util.go
+++ b/deepfence_agent/tools/apache/scope/cache/util.go
@@ -6,6 +6,7 @@ import (
 	"fmt"
 	"github.com/ugorji/go/codec"
 	"github.com/weaveworks/scope/report"
+	"io"
 	"os"
 	"os/exec"
 	"reflect"
@@ -175,6 +176,14 @@ func CodecEncode(data interface{}) ([]byte, error) {
 		return nil, err
 	}
 	return buf.Bytes(), nil
+}
+
+func CodecDecode(reader io.Reader, data interface{}) error {
+	err := codec.NewDecoder(reader, &codec.JsonHandle{}).Decode(data)
+	if err != nil {
+		return err
+	}
+	return nil
 }
 
 func (topologyOptions *TopologyOptions) TopologyOptionsValidate() {

